bgzf/writer.go

```diff
@@ -0,0 +1,111 @@
+package bgzf
+
+import (
+	"compress/gzip"
+	"encoding/binary"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"log"
+	"math"
+)
+
+// magicHexEOF is a particular empty bgzf block that marks the true EOF.
+var magicEOF []byte = makeMagicEOF()
+
+// Writer moves data -> compressor -> zipBlock -> file writer
+type Writer struct {
+	w          io.Writer
+	compressor *gzip.Writer
+	zipBlock   *Block
+}
+
+// NewWriter creates a bgzf writer from any input writer.
+func NewWriter(w io.Writer) Writer {
+	var zw Writer
+	zw.w = w
+	zw.zipBlock = NewBlock()
+	zw.compressor = gzip.NewWriter(zw.zipBlock)
+	return zw
+}
+
+// Write input bytes as a single bgzf block.
+func (w Writer) Write(p []byte) (n int, err error) {
+	w.compressor.Reset(w.zipBlock)
+
+	// ******
+	// TODO better way around this?
+	// The gzip writer tries to write a header smartly at the time of
+	// data write, however bgzf requires custom header fields which
+	// require knowing the compressed size of the block, which can only
+	// be determined after compression. Therefore we cannot use the
+	// default gzip header behavior. The code below performs an empty
+	// write to make the gzip writer 'think' it has written a header.
+	// We then trash the default header and write a custom header below.
+	_, err = w.compressor.Write([]byte{}) // trash gzip default header
+	if err != nil {
+		log.Panic(err)
+	}
+	w.zipBlock.Reset() // reset written trashed header
+	// end of code block for discarding default gzip header
+	// ******
+
+	n, err = w.compressor.Write(p) // compress p into write buffer
+	if n != len(p) || err != nil {
+		fmt.Printf("input %d bytes, wrote %d bytes\n", len(p), n)
+		log.Panic(err)
+	}
+	err = w.compressor.Close()
+	if err != nil {
+		log.Panic(err)
+	}
+
+	w.writeHeader(w.zipBlock.Len())        // write block header to file based on compressed size
+	_, err = w.w.Write(w.zipBlock.Bytes()) // write buffer to file
+	return
+}
+
+// Close bgzf writer and add the magic EOF marker to the end of the file.
+func (w Writer) Close() error {
+	err := w.compressor.Close()
+	if err != nil {
+		log.Panic(err)
+	}
+	_, err = w.w.Write(magicEOF)
+	return err
+}
+
+// writeHeader for custom bgzf header.
+func (w Writer) writeHeader(compSize int) {
+	var header [18]byte
+	header[0] = 31  // gzip ID 1
+	header[1] = 139 // gzip ID 2
+	header[2] = 8   // gzip compression method
+	header[3] = 4   // gzip flags
+	// header[4:8] = mod time uint32
```

Comment:
Does this mean that headers 4 through 8 are blank or skipped? I'm trying ti visualize this in my head.

Reply:
This means that 4-8 are set to 0 (the default value). I initialize the array to have a specific size where everything is set to zero by default. If I want any values to stay zero, I can just ignore them. I just leave the comment here for clarity.